atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/KvsSweepQueueWriter.java

```diff
@@ -16,29 +16,45 @@
 
 package com.palantir.atlasdb.sweep.queue;
 
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
 import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.ColumnRangeSelection;
 import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.RangeRequest;
+import com.palantir.atlasdb.keyvalue.api.RowColumnRangeIterator;
 import com.palantir.atlasdb.keyvalue.api.TableReference;
 
 public abstract class KvsSweepQueueWriter implements SweepQueueWriter {
     private final KeyValueService kvs;
     private final TableReference tableRef;
+    private final WriteInfoPartitioner partitioner;
 
-    public KvsSweepQueueWriter(KeyValueService kvs, TableReference tableRef) {
+    public KvsSweepQueueWriter(KeyValueService kvs, TableReference tableRef, WriteInfoPartitioner partitioner) {
         this.kvs = kvs;
         this.tableRef = tableRef;
+        this.partitioner = partitioner;
     }
 
     @Override
-    public void enqueue(List<WriteInfo> writes) {
-        Map<Cell, byte[]> batchedWrites = batchWrites(writes);
-        if (!batchedWrites.isEmpty()) {
-            kvs.put(tableRef, batchWrites(writes), 0L);
+    public void enqueue(List<WriteInfo> allWrites) {
+        Map<Cell, byte[]> cellsToWrite = new HashMap<>();
+        Map<PartitionInfo, List<WriteInfo>> partitionedWrites = partitioner.filterAndPartition(allWrites);
+        partitionedWrites.forEach((partitionInfo, writes) -> populateCells(partitionInfo, writes, cellsToWrite));
```

Comment:
Note: It's kind of janky that `cellsToWrite` gets passed down through many levels of the call stack before someone calls `put`, rather than having `populateCells` fetch the cells which are then `putAll` here. I guess your way works out more efficient, as we don't need to create tons of mini-maps but just have this one map to rule them all?

Reply:
This is to keep only the write with the maximum timestamp for each cell in memory at a time. Given that we will also need to do lookups to determine which writes were actually committed, the approach you mention makes more sense. Will implement that all in one PR